File: frontend/services/mentorshipService.ts
import { getAuthToken } from "./authService";

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";

export interface MentorshipRequestDto {
  id: string;
  mentorId: string;
  mentorName: string;
  mentorProfilePicture?: string;
  menteeId: string;
  menteeName: string;
  menteeProfilePicture?: string;
  status: string;
  message?: string;
  sessionDate?: string;
  feedback?: string;
  rating?: number;
}

export interface MentorshipRequestCreateDto {
  mentorId: string;
  message?: string;
  sessionDate?: string;
}

async function authFetch(url: string, options: RequestInit = {}) {
  const token = getAuthToken();
  if (!token) throw new Error("No authentication token found.");
  const res = await fetch(url, {
    ...options,
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json", ...options.headers },
  });
  if (!res.ok) throw new Error(await res.text() || "Request failed");
  return res;
}

export async function registerAsMentor(params: { skills: string; experience: string; company: string; availability: string; meetingMode: string; helpAreas: string }): Promise<void> {
  const qs = new URLSearchParams(params).toString();
  await authFetch(`${API_BASE}/api/mentorship/register?${qs}`, { method: "POST" });
}

export async function requestMentorship(dto: MentorshipRequestCreateDto): Promise<MentorshipRequestDto> {
  const res = await authFetch(`${API_BASE}/api/mentorship/request`, { method: "POST", body: JSON.stringify(dto) });
  return res.json();
}

export async function acceptMentorship(id: string): Promise<MentorshipRequestDto> {
  const res = await authFetch(`${API_BASE}/api/mentorship/requests/${id}/accept`, { method: "POST" });
  return res.json();
}

export async function rejectMentorship(id: string): Promise<MentorshipRequestDto> {
  const res = await authFetch(`${API_BASE}/api/mentorship/requests/${id}/reject`, { method: "POST" });
  return res.json();
}

export async function scheduleSession(id: string, sessionDate: string): Promise<MentorshipRequestDto> {
  const res = await authFetch(`${API_BASE}/api/mentorship/requests/${id}/schedule?sessionDate=${encodeURIComponent(sessionDate)}`, { method: "POST" });
  return res.json();
}

export async function submitFeedback(id: string, feedback: string, rating: number): Promise<MentorshipRequestDto> {
  const res = await authFetch(`${API_BASE}/api/mentorship/requests/${id}/feedback?feedback=${encodeURIComponent(feedback)}&rating=${rating}`, { method: "POST" });
  return res.json();
}

export async function getRequestsAsMentor(): Promise<MentorshipRequestDto[]> {
  const res = await authFetch(`${API_BASE}/api/mentorship/requests/mentor`);
  return res.json();
}

export async function getRequestsAsMentee(): Promise<MentorshipRequestDto[]> {
  const res = await authFetch(`${API_BASE}/api/mentorship/requests/mentee`);
  return res.json();
}

export async function getMentors(): Promise<any[]> {
  const res = await authFetch(`${API_BASE}/api/mentorship/mentors`);
  return res.json();
}

File: backend/src/main/java/com/alumnihub/service/MentorshipService.java
package com.alumnihub.service;

import com.alumnihub.dto.MentorshipRequestCreateDto;
import com.alumnihub.dto.MentorshipRequestDto;
import com.alumnihub.dto.UserDto;
import com.alumnihub.entity.MentorshipRequest;
import com.alumnihub.entity.User;
import com.alumnihub.repository.MentorshipRequestRepository;
import com.alumnihub.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class MentorshipService {

    private final MentorshipRequestRepository mentorshipRequestRepository;
    private final UserRepository userRepository;

    @Transactional
    public void registerAsMentor(String userEmail, String skills, String experience, String company, String availability, String meetingMode, String helpAreas) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        user.setMentorSkills(skills);
        user.setMentorExperience(experience);
        user.setMentorCompany(company);
        user.setMentorAvailability(availability);
        user.setMentorMeetingMode(meetingMode);
        user.setMentorHelpAreas(helpAreas);

        // Add Mentor badge if not already present
        String currentBadges = user.getBadges();
        if (currentBadges == null || currentBadges.trim().isEmpty()) {
            user.setBadges("Mentor");
        } else if (!currentBadges.contains("Mentor")) {
            user.setBadges(currentBadges + ",Mentor");
        }

        userRepository.save(user);
    }

    @Transactional
    public MentorshipRequestDto requestMentorship(String menteeEmail, MentorshipRequestCreateDto createDto) {
        User mentee = userRepository.findByEmail(menteeEmail)
                .orElseThrow(() -> new IllegalArgumentException("Mentee not found: " + menteeEmail));

        User mentor = userRepository.findById(createDto.getMentorId())
                .orElseThrow(() -> new IllegalArgumentException("Mentor not found: " + createDto.getMentorId()));

        MentorshipRequest request = MentorshipRequest.builder()
                .mentee(mentee)
                .mentor(mentor)
                .message(createDto.getMessage())
                .sessionDate(createDto.getSessionDate())
                .status("PENDING")
                .build();

        MentorshipRequest saved = mentorshipRequestRepository.save(request);
        return convertToDto(saved);
    }

    @Transactional
    public MentorshipRequestDto acceptMentorship(String mentorEmail, UUID requestId) {
        MentorshipRequest request = mentorshipRequestRepository.findById(requestId)
                .orElseThrow(() -> new IllegalArgumentException("Request not found: " + requestId));

        if (!request.getMentor().getEmail().equalsIgnoreCase(mentorEmail)) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized to accept this request.");
        }

        request.setStatus("ACCEPTED");
        return convertToDto(mentorshipRequestRepository.save(request));
    }

    @Transactional
    public MentorshipRequestDto rejectMentorship(String mentorEmail, UUID requestId) {
        MentorshipRequest request = mentorshipRequestRepository.findById(requestId)
                .orElseThrow(() -> new IllegalArgumentException("Request not found: " + requestId));

        if (!request.getMentor().getEmail().equalsIgnoreCase(mentorEmail)) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized to reject this request.");
        }

        request.setStatus("REJECTED");
        return convertToDto(mentorshipRequestRepository.save(request));
    }

    @Transactional
    public MentorshipRequestDto scheduleSession(String userEmail, UUID requestId, LocalDateTime sessionDate) {
        MentorshipRequest request = mentorshipRequestRepository.findById(requestId)
                .orElseThrow(() -> new IllegalArgumentException("Request not found: " + requestId));

        if (!request.getMentor().getEmail().equalsIgnoreCase(userEmail) && !request.getMentee().getEmail().equalsIgnoreCase(userEmail)) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized.");
        }

        request.setSessionDate(sessionDate);
        return convertToDto(mentorshipRequestRepository.save(request));
    }

    @Transactional
    public MentorshipRequestDto submitFeedback(String menteeEmail, UUID requestId, String feedback, Integer rating) {
        MentorshipRequest request = mentorshipRequestRepository.findById(requestId)
                .orElseThrow(() -> new IllegalArgumentException("Request not found: " + requestId));

        if (!request.getMentee().getEmail().equalsIgnoreCase(menteeEmail)) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized to submit feedback.");
        }

        request.setFeedback(feedback);
        request.setRating(rating);
        return convertToDto(mentorshipRequestRepository.save(request));
    }

    public List<MentorshipRequestDto> getMentorshipRequests(String userEmail, boolean asMentor) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        List<MentorshipRequest> requests = asMentor
                ? mentorshipRequestRepository.findAllByMentor(user)
                : mentorshipRequestRepository.findAllByMentee(user);

        return requests.stream().map(this::convertToDto).collect(Collectors.toList());
    }

    public List<User> getActiveMentors() {
        return userRepository.findAll().stream()
                .filter(u -> u.getBadges() != null && u.getBadges().contains("Mentor"))
                .collect(Collectors.toList());
    }

    private MentorshipRequestDto convertToDto(MentorshipRequest r) {
        return MentorshipRequestDto.builder()
                .id(r.getId())
                .mentorId(r.getMentor().getId())
                .mentorName(r.getMentor().getFullName())
                .mentorProfilePicture(r.getMentor().getProfilePictureUrl())
                .menteeId(r.getMentee().getId())
                .menteeName(r.getMentee().getFullName())
                .menteeProfilePicture(r.getMentee().getProfilePictureUrl())
                .status(r.getStatus())
                .message(r.getMessage())
                .sessionDate(r.getSessionDate())
                .feedback(r.getFeedback())
                .rating(r.getRating())
                .build();
    }
}
